h2/src/main/org/h2/engine/Session.java

```diff
@@ -681,16 +673,27 @@ private void markUsedTablesAsUpdated() {
     }
 
     private void analyzeTables() {
-        // take a local copy and clear because in rare cases we can call
-        // back into markTableForAnalyze while iterating here
-        HashSet<Table> tablesToAnalyzeLocal = tablesToAnalyze;
-        tablesToAnalyze = null;
-        int rowCount = getDatabase().getSettings().analyzeSample / 10;
-        for (Table table : tablesToAnalyzeLocal) {
-            Analyze.analyzeTable(this, table, rowCount, false);
-        }
-        // analyze can lock the meta
-        database.unlockMeta(this);
+        // On rare occasions it can be called concurrently (i.e. from close())
+        // whithout proper locking, but instead of oversynchronizing
+        // we just skip this optional operation in such case
+        if (Thread.holdsLock(database.isPersistent() ? this : database)
+                && tablesToAnalyze != null) {
```

Comment:
I think cheaper `tablesToAnalyze` should be checked first.

And I don't understand why `isPersistent()` is used instead of `isMVStore()`.

Reply:
Oops, Thanks @katzyn, that's what peers are for, especially on Thanksgiving night!

Reply:
And yes, tablesToAnalyze is cheaper to check, but in absence of synchronization it would be a check of thread-local memory and can lead to NPE. As I wrote this sentence, I've come to realize that although we would check garbage in the first part (ablesToAnalyze != null), second part of condition will prevent NPE from happening anyway, so you are right.